qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_tof_powder_dataset.py

```diff
@@ -0,0 +1,186 @@
+                   for key, value in field_items]
+    return ",\n".join(string_list)
+
+
+def _get_path_string(dataset, sample_name):
+    path = dataset[sample_name].pop('path')
+    return f"'path': '{path}',\n"
+
+
+def _get_sample_string(sample_name, llens):
+    return f"'{sample_name:s}': {{".rjust(llens)
+
+
+def _get_datapath(entry, path):
+    proposal = get_proposal_from_filename(entry['filename'],
+                                          entry['file_number'])
+    return os.path.join(path, proposal)
+
+
+def _convert_list_to_range(dataset):
+    for sample_typ, det_rots in dataset.items():
+        for det_rot, file_numbers in det_rots.items():
+            if det_rot != 'path':
+                dataset[sample_typ][det_rot] = list_to_range(file_numbers)
+    return dataset
+
+
+def _get_closest_bank_key(banks, det_rot):
+    rounding_limit = 0.05
+    compare = det_rot
+    for compare in [x for x in banks.keys() if x != 'path']:
+        if abs(compare - det_rot) < rounding_limit:
+            break
+        compare = det_rot
+    return compare
+
+
+def _create_new_datatype(dataset, datatype, det_rot, entry, path):
+    dataset[datatype] = {
+        det_rot: [entry['file_number']],
+        'path': _get_datapath(entry, path)
+    }
+
+
+def _add_or_create_filelist(dataset, datatype, det_rot, entry):
+    compare = _get_closest_bank_key(dataset[datatype], det_rot)
+    fn_list = dataset[datatype].get(compare, None)
+    if fn_list is not None:
+        fn_list.append(entry['file_number'])
+    else:
+        dataset[datatype][compare] = [entry['file_number']]
+
+
+class DNSTofDataset(ObjectDict):
+    """
+    Class for storing data of a multiple dns datafiles.
+    This is a dictionary, but can also be accessed like attributes.
+    """
+    def __init__(self, data, path, is_sample=True):
+        super().__init__()
+        self['is_sample'] = is_sample
+        self['banks'] = get_bank_positions(data)
+        self['data_dic'] = self.create_dataset(data, path)
+
+    def format_dataset(self):
+        """
+        Formatting the dictionary to a nicely indented string.
+        """
+
+        dataset = self.data_dic
+        llens = _get_max_key_length(dataset)
+        dataset_string = '{\n'
+        for sample_name, fields in dataset.items():
+            dataset_string += _get_sample_string(sample_name, llens)
+            dataset_string += _get_path_string(dataset, sample_name)
+            dataset_string += _get_field_string(fields, llens)
+            dataset_string += "},\n"
+        dataset_string += "}"
+        return dataset_string
+
+    def _get_nb_banks(self, sample_type=None):
+        dic = self.data_dic.get(sample_type, 0)
+        if dic:
+            return len(dic.keys()) - 1
+        return 0
+
+    def get_vana_filename(self):
+        vana_filename = [x for x in self.data_dic.keys() if '_vana' in x]
+        if len(vana_filename) == 0:
+            vana_filename = ''
+        elif len(vana_filename) > 1:
```

Comment:
Am I right to assume that there should never be more than one Vanadium filename? if so should there be a error message if two are found?

Reply:
Good point, thanks! The error message will be modified to mention that only one vanadium "filename" can be selected if the vanadium correction is chosen to be calculated. Actually, the naming of this variable (and the method where this variable is defined) is misleading, because it describes rather the name of the scan than the filename. Several vanadium files can be chosen, but they have to belong to the same scan. So this method and the variable will be refactored